Firestore/core/src/firebase/firestore/remote/remote_store.mm

```diff
@@ -291,6 +289,187 @@
+    last_batch_id_retrieved = batch.batchID;
+  }
+
+  if (ShouldStartWriteStream()) {
+    StartWriteStream();
+  }
+}
+
+bool RemoteStore::CanAddToWritePipeline() const {
+  return CanUseNetwork() && write_pipeline_.size() < kMaxPendingWrites;
+}
+
+void RemoteStore::AddToWritePipeline(FSTMutationBatch* batch) {
+  HARD_ASSERT(CanAddToWritePipeline(),
+              "AddToWritePipeline called when pipeline is full");
+
+  write_pipeline_.push_back(batch);
+
+  if (write_stream_->IsOpen() && write_stream_->handshake_complete()) {
+    write_stream_->WriteMutations(batch.mutations);
+  }
+}
+
+bool RemoteStore::ShouldStartWriteStream() const {
+  return CanUseNetwork() && !write_stream_->IsStarted() &&
+         !write_pipeline_.empty();
+}
+
+void RemoteStore::StartWriteStream() {
+  HARD_ASSERT(ShouldStartWriteStream(), "StartWriteStream called when "
+                                        "ShouldStartWriteStream is false.");
+  write_stream_->Start();
+}
+
+void RemoteStore::OnWriteStreamOpen() {
+  write_stream_->WriteHandshake();
+}
+
+void RemoteStore::OnWriteStreamHandshakeComplete() {
+  // Record the stream token.
+  [local_store_ setLastStreamToken:write_stream_->GetLastStreamToken()];
+
+  // Send the write pipeline now that the stream is established.
+  for (FSTMutationBatch* write : write_pipeline_) {
+    write_stream_->WriteMutations(write.mutations);
+  }
+}
+
+void RemoteStore::OnWriteStreamMutationResult(
+    SnapshotVersion commit_version,
+    std::vector<FSTMutationResult*> mutation_results) {
+  // This is a response to a write containing mutations and should be correlated
+  // to the first write in our write pipeline.
+  HARD_ASSERT(!write_pipeline_.empty(), "Got result for empty write pipeline");
+
+  FSTMutationBatch* batch = write_pipeline_.front();
+  write_pipeline_.erase(write_pipeline_.begin());
+
+  FSTMutationBatchResult* batchResult = [FSTMutationBatchResult
+      resultWithBatch:batch
+        commitVersion:commit_version
+      mutationResults:std::move(mutation_results)
+          streamToken:write_stream_->GetLastStreamToken()];
+  [sync_engine_ applySuccessfulWriteWithResult:batchResult];
+
+  // It's possible that with the completion of this mutation another slot has
+  // freed up.
+  FillWritePipeline();
+}
+
+/**
+ * Handles the closing of the StreamingWrite RPC, either because of an error or
+ * because the RPC has been terminated by the client or the server.
+ */
+void RemoteStore::OnWriteStreamClose(const Status& status) {
+  if (status.ok()) {
+    // Graceful stop (due to Stop() or idle timeout). Make sure that's
+    // desirable.
+    HARD_ASSERT(!ShouldStartWriteStream(),
+                "Write stream was stopped gracefully while still needed.");
+  }
+
+  // If the write stream closed due to an error, invoke the error callbacks if
+  // there are pending writes.
+  if (!status.ok() && !write_pipeline_.empty()) {
+    if (write_stream_->handshake_complete()) {
```

Comment:
Android has the following TODO here:
```
TODO: handle UNAUTHENTICATED status, see go/firestore-client-errors
```
I don't know if it's relevant.

Reply:
I decided to just add the same text for consistency. The doc is 2 years old at this point, so I'm not sure we still want to implement the error handling as described.